python_modules/libraries/dagster-tableau/dagster_tableau/resources.py

```diff
@@ -0,0 +1,154 @@
+        response = self.sign_in()
+        self._api_token = response["credentials"]["token"]
+        self._site_id = response["credentials"]["site"]["id"]
+
+    def teardown_after_execution(self, context: InitResourceContext) -> None:
+        # Sign out after execution
+        self.sign_out()
+        self._api_token = None
+        self._site_id = None
+
+    @property
+    @abstractmethod
+    def api_base_url(self) -> str:
+        raise NotImplementedError()
+
+    def _fetch_json(
+        self,
+        endpoint: str,
+        data: Optional[Dict[str, Any]] = None,
+        method: str = "GET",
+        with_auth_header: bool = True,
+    ) -> Dict[str, Any]:
+        """Fetch JSON data from the Tableau API. Raises an exception if the request fails.
+
+        Args:
+            endpoint (str): The API endpoint to fetch data from.
+            data (Optional[Dict[str, Any]]): JSON-formatted data string to be included in the request.
+            method (str): The HTTP method to use for the request.
+            with_auth_header (bool): Whether to add X-Tableau-Auth header to the request. Enabled by default.
+
+        Returns:
+            Dict[str, Any]: The JSON data returned from the API.
+        """
+        response = self._make_request(
+            endpoint=endpoint, data=data, method=method, with_auth_header=with_auth_header
+        )
+        return response.json()
+
+    def _make_request(
+        self,
+        endpoint: str,
+        data: Optional[Dict[str, Any]] = None,
+        method: str = "GET",
+        with_auth_header: bool = True,
+    ) -> requests.Response:
+        headers = {
+            "accept": "application/json",
+            "content-type": "application/json",
+        }
+        if with_auth_header:
+            headers["X-tableau-auth"] = self._api_token
+        request_args: Dict[str, Any] = dict(
+            method=method,
+            url=f"{self.api_base_url}/{endpoint}",
+            headers=headers,
+        )
+        if data:
+            request_args["json"] = data
+        response = requests.request(**request_args)
+        response.raise_for_status()
+        return response
+
+    @cached_method
+    def get_workbooks(self) -> Dict[str, Any]:
+        """Fetches a list of all Tableau workbooks in the workspace."""
+        return self._fetch_json(self._with_site_id("workbooks"))
+
+    @cached_method
+    def get_workbook(self, workbook_id) -> Dict[str, Any]:
+        """Fetches information, including views and tags, for a given workbook."""
+        return self._fetch_json(self._with_site_id(f"workbooks/{workbook_id}"))
+
+    @cached_method
+    def get_workbook_data_sources(
+        self,
+        workbook_id: str,
+    ) -> Dict[str, Any]:
+        """Fetches a list of all data sources for a given workbook."""
+        return self._fetch_json(self._with_site_id(f"workbooks/{workbook_id}/connections"))
+
+    def sign_in(self):
```

Comment:
Nitpick: add return type annotation

Reply:
Done in [12535aa](https://github.com/dagster-io/dagster/pull/24233/commits/12535aa90ff10d1ca0264b27a31bbb12ae67a803)